`clangxx` does not include `gxx_impl_linux-64` as dependency

### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

When installing `clangxx=16` (not sure if it's specific to 16), `gxx_impl_linux-64` is not pulled in automatically as a dependency, leading to this issue when trying to compile a simple cpp program:

```
[17:52:44 | last: 0s] (  1) | ~
igozali@host $ clang foo.cpp
foo.cpp:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.
```

Receipt of `mamba install`:
```
[17:50:03 | last: 1s] (  1) | ~
igozali@host $ mamba create -n clangxx16 -c conda-forge clangxx=16

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['clangxx=16']

conda-forge/noarch                                  14.2MB @   6.5MB/s  2.3s
conda-forge/linux-64                                35.1MB @   7.5MB/s  5.0s
Transaction

  Prefix: /home/igozali/.conda/envs/clangxx16

  Updating specs:

   - clangxx=16


  Package                    Version  Build               Channel                    Size
───────────────────────────────────────────────────────────────────────────────────────────
  Install:
───────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex                0.1  conda_forge         conda-forge/linux-64     Cached
  + _openmp_mutex                4.5  2_gnu               conda-forge/linux-64     Cached
  + binutils_impl_linux-64      2.40  hf600244_0          conda-forge/linux-64     Cached
  + clang                     16.0.6  ha0738ec_1          conda-forge/linux-64     Cached
  + clang-16                  16.0.6  default_h1cdf331_1  conda-forge/linux-64     Cached
  + clangxx                   16.0.6  default_h1cdf331_1  conda-forge/linux-64     Cached
  + gcc_impl_linux-64         13.2.0  h338b0a0_2          conda-forge/linux-64     Cached
  + icu                         73.2  h59595ed_0          conda-forge/linux-64     Cached
  + kernel-headers_linux-64   2.6.32  he073ed8_16         conda-forge/noarch       Cached
  + ld_impl_linux-64            2.40  h41732ed_0          conda-forge/linux-64     Cached
  + libclang-cpp16            16.0.6  default_h1cdf331_1  conda-forge/linux-64     Cached
  + libgcc-devel_linux-64     13.2.0  ha9c7c90_2          conda-forge/linux-64     Cached
  + libgcc-ng                 13.2.0  h807b86a_2          conda-forge/linux-64     Cached
  + libgomp                   13.2.0  h807b86a_2          conda-forge/linux-64     Cached
  + libiconv                    1.17  h166bdaf_0          conda-forge/linux-64     Cached
  + libllvm16                 16.0.6  h5cf9203_2          conda-forge/linux-64     Cached
  + libsanitizer              13.2.0  h7e041cc_2          conda-forge/linux-64     Cached
  + libstdcxx-ng              13.2.0  h7e041cc_2          conda-forge/linux-64     Cached
  + libxml2                   2.11.5  h232c23b_1          conda-forge/linux-64     Cached
  + libzlib                   1.2.13  hd590300_5          conda-forge/linux-64     Cached
  + sysroot_linux-64            2.12  he073ed8_16         conda-forge/noarch       Cached
  + xz                         5.2.6  h166bdaf_0          conda-forge/linux-64     Cached
  + zstd                       1.5.5  hfc55251_0          conda-forge/linux-64     Cached

  Summary:

  Install: 23 packages

  Total download: 0 B

───────────────────────────────────────────────────────────────────────────────────────────
```

### Installed packages

```shell
$ mamba list
# packages in environment at /home/igozali/.conda/envs/clangxx16:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
binutils_impl_linux-64    2.40                 hf600244_0    conda-forge
clang                     16.0.6               ha0738ec_1    conda-forge
clang-16                  16.0.6          default_h1cdf331_1    conda-forge
clangxx                   16.0.6          default_h1cdf331_1    conda-forge
gcc_impl_linux-64         13.2.0               h338b0a0_2    conda-forge
icu                       73.2                 h59595ed_0    conda-forge
kernel-headers_linux-64   2.6.32              he073ed8_16    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libclang-cpp16            16.0.6          default_h1cdf331_1    conda-forge
libgcc-devel_linux-64     13.2.0               ha9c7c90_2    conda-forge
libgcc-ng                 13.2.0               h807b86a_2    conda-forge
libgomp                   13.2.0               h807b86a_2    conda-forge
libiconv                  1.17                 h166bdaf_0    conda-forge
libllvm16                 16.0.6               h5cf9203_2    conda-forge
libsanitizer              13.2.0               h7e041cc_2    conda-forge
libstdcxx-ng              13.2.0               h7e041cc_2    conda-forge
libxml2                   2.11.5               h232c23b_1    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
sysroot_linux-64          2.12                he073ed8_16    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
zstd                      1.5.5                hfc55251_0    conda-forge
```


### Environment info

```shell
$ mamba info

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : clangxx16
    active env location : /home/igozali/.conda/envs/clangxx16
            shell level : 1
       user config file : /home/igozali/.condarc
 populated config files : /conda-base/miniconda-py38_4.8.3/.condarc
          conda version : 23.3.1
    conda-build version : not installed
         python version : 3.8.16.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.31=0
                          __linux=5.4.0=0
                          __unix=0=0
       base environment : :/conda-base/miniconda-py38_4.8.3  (writable)
      conda av data dir : /conda-base/miniconda-py38_4.8.3/etc/conda
  conda av metadata url : None
           channel URLs :
          package cache : /home/igozali/.conda/pkgs
                          /conda-base/miniconda-py38_4.8.3/pkgs
       envs directories : /home/igozali/.conda/envs
                          /conda-envs
                          /conda-base/miniconda-py38_4.8.3/envs
               platform : linux-64
             user-agent : conda/23.3.1 requests/2.29.0 CPython/3.8.16 Linux/5.4.0-42-generic ubuntu/20.04 glibc/2.31
                UID:GID : 1000:100
             netrc file : /home/igozali/.netrc
           offline mode : False
```

--- a/.azure-pipelines/azure-pipelines-linux.yml
+++ b/.azure-pipelines/azure-pipelines-linux.yml
@@ -33,6 +33,9 @@ jobs:
 
   - script: |
         export CI=azure
+        export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
+        export remote_url=$(Build.Repository.Uri)
+        export sha=$(Build.SourceVersion)
         export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
         export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})
         if [[ "${BUILD_REASON:-}" == "PullRequest" ]]; then

--- a/.azure-pipelines/azure-pipelines-osx.yml
+++ b/.azure-pipelines/azure-pipelines-osx.yml
@@ -20,6 +20,9 @@ jobs:
   # TODO: Fast finish on azure pipelines?
   - script: |
       export CI=azure
+      export flow_run_id=azure_$(Build.BuildNumber).$(System.JobAttempt)
+      export remote_url=$(Build.Repository.Uri)
+      export sha=$(Build.SourceVersion)
       export OSX_FORCE_SDK_DOWNLOAD="1"
       export GIT_BRANCH=$BUILD_SOURCEBRANCHNAME
       export FEEDSTOCK_NAME=$(basename ${BUILD_REPOSITORY_NAME})

--- a/.azure-pipelines/azure-pipelines-win.yml
+++ b/.azure-pipelines/azure-pipelines-win.yml
@@ -42,6 +42,9 @@ jobs:
         PYTHONUNBUFFERED: 1
         CONFIG: $(CONFIG)
         CI: azure
+        flow_run_id: azure_$(Build.BuildNumber).$(System.JobAttempt)
+        remote_url: $(Build.Repository.Uri)
+        sha: $(Build.SourceVersion)
         UPLOAD_PACKAGES: $(UPLOAD_PACKAGES)
         UPLOAD_TEMP: $(UPLOAD_TEMP)
         BINSTAR_TOKEN: $(BINSTAR_TOKEN)

--- a/.scripts/build_steps.sh
+++ b/.scripts/build_steps.sh
@@ -28,13 +28,15 @@ conda-build:
 pkgs_dirs:
   - ${FEEDSTOCK_ROOT}/build_artifacts/pkg_cache
   - /opt/conda/pkgs
+solver: libmamba
 
 CONDARC
+export CONDA_LIBMAMBA_SOLVER_NO_CHANNELS_FROM_INSTALLED=1
 
 mamba install --update-specs --yes --quiet --channel conda-forge --strict-channel-priority \
-    pip mamba conda-build boa conda-forge-ci-setup=3
+    pip mamba conda-build conda-forge-ci-setup=4
 mamba update --update-specs --yes --quiet --channel conda-forge --strict-channel-priority \
-    pip mamba conda-build boa conda-forge-ci-setup=3
+    pip mamba conda-build conda-forge-ci-setup=4
 
 # set up the condarc
 setup_conda_rc "${FEEDSTOCK_ROOT}" "${RECIPE_ROOT}" "${CONFIG_FILE}"
@@ -55,6 +57,12 @@ if [[ -f "${FEEDSTOCK_ROOT}/LICENSE.txt" ]]; then
   cp "${FEEDSTOCK_ROOT}/LICENSE.txt" "${RECIPE_ROOT}/recipe-scripts-license.txt"
 fi
 
+if [[ "${sha:-}" == "" ]]; then
+  pushd ${FEEDSTOCK_ROOT}
+  sha=$(git rev-parse HEAD)
+  popd
+fi
+
 if [[ "${BUILD_WITH_CONDA_DEBUG:-0}" == 1 ]]; then
     if [[ "x${BUILD_OUTPUT_ID:-}" != "x" ]]; then
         EXTRA_CB_OPTIONS="${EXTRA_CB_OPTIONS:-} --output-id ${BUILD_OUTPUT_ID}"
@@ -66,9 +74,10 @@ if [[ "${BUILD_WITH_CONDA_DEBUG:-0}" == 1 ]]; then
     # Drop into an interactive shell
     /bin/bash
 else
-    conda mambabuild "${RECIPE_ROOT}" -m "${CI_SUPPORT}/${CONFIG}.yaml" \
+    conda build "${RECIPE_ROOT}" -m "${CI_SUPPORT}/${CONFIG}.yaml" \
         --suppress-variables ${EXTRA_CB_OPTIONS:-} \
-        --clobber-file "${CI_SUPPORT}/clobber_${CONFIG}.yaml"
+        --clobber-file "${CI_SUPPORT}/clobber_${CONFIG}.yaml" \
+        --extra-meta flow_run_id="${flow_run_id:-}" remote_url="${remote_url:-}" sha="${sha:-}"
     ( startgroup "Validating outputs" ) 2> /dev/null
 
     validate_recipe_outputs "${FEEDSTOCK_NAME}"

--- a/.scripts/run_docker_build.sh
+++ b/.scripts/run_docker_build.sh
@@ -91,6 +91,9 @@ docker run ${DOCKER_RUN_ARGS} \
            -e CPU_COUNT \
            -e BUILD_WITH_CONDA_DEBUG \
            -e BUILD_OUTPUT_ID \
+           -e flow_run_id \
+           -e remote_url \
+           -e sha \
            -e BINSTAR_TOKEN \
            -e FEEDSTOCK_TOKEN \
            -e STAGING_BINSTAR_TOKEN \

--- a/.scripts/run_osx_build.sh
+++ b/.scripts/run_osx_build.sh
@@ -22,11 +22,13 @@ bash $MINIFORGE_FILE -b -p ${MINIFORGE_HOME}
 
 source ${MINIFORGE_HOME}/etc/profile.d/conda.sh
 conda activate base
+export CONDA_SOLVER="libmamba"
+export CONDA_LIBMAMBA_SOLVER_NO_CHANNELS_FROM_INSTALLED=1
 
 mamba install --update-specs --quiet --yes --channel conda-forge --strict-channel-priority \
-    pip mamba conda-build boa conda-forge-ci-setup=3
+    pip mamba conda-build conda-forge-ci-setup=4
 mamba update --update-specs --yes --quiet --channel conda-forge --strict-channel-priority \
-    pip mamba conda-build boa conda-forge-ci-setup=3
+    pip mamba conda-build conda-forge-ci-setup=4
 
 
 
@@ -43,6 +45,10 @@ if [[ "${CI:-}" != "" ]]; then
   /usr/bin/sudo -k
 else
   echo -e "\n\nNot mangling homebrew as we are not running in CI"
+fi
+
+if [[ "${sha:-}" == "" ]]; then
+  sha=$(git rev-parse HEAD)
 fi
 
 echo -e "\n\nRunning the build setup script."
@@ -75,9 +81,10 @@ else
         EXTRA_CB_OPTIONS="${EXTRA_CB_OPTIONS:-} --no-test"
     fi
 
-    conda mambabuild ./recipe -m ./.ci_support/${CONFIG}.yaml \
+    conda build ./recipe -m ./.ci_support/${CONFIG}.yaml \
         --suppress-variables ${EXTRA_CB_OPTIONS:-} \
-        --clobber-file ./.ci_support/clobber_${CONFIG}.yaml
+        --clobber-file ./.ci_support/clobber_${CONFIG}.yaml \
+        --extra-meta flow_run_id="$flow_run_id" remote_url="$remote_url" sha="$sha"
     ( startgroup "Validating outputs" ) 2> /dev/null
 
     validate_recipe_outputs "${FEEDSTOCK_NAME}"

--- a/.scripts/run_win_build.bat
+++ b/.scripts/run_win_build.bat
@@ -17,10 +17,14 @@ call :start_group "Configuring conda"
 
 :: Activate the base conda environment
 call activate base
+:: Configure the solver
+set "CONDA_SOLVER=libmamba"
+if !errorlevel! neq 0 exit /b !errorlevel!
+set "CONDA_LIBMAMBA_SOLVER_NO_CHANNELS_FROM_INSTALLED=1"
 
 :: Provision the necessary dependencies to build the recipe later
 echo Installing dependencies
-mamba.exe install "python=3.10" pip mamba conda-build boa conda-forge-ci-setup=3 -c conda-forge --strict-channel-priority --yes
+mamba.exe install "python=3.10" pip mamba conda-build conda-forge-ci-setup=4 -c conda-forge --strict-channel-priority --yes
 if !errorlevel! neq 0 exit /b !errorlevel!
 
 :: Set basic configuration
@@ -41,11 +45,15 @@ if NOT [%HOST_PLATFORM%] == [%BUILD_PLATFORM%] (
     set "EXTRA_CB_OPTIONS=%EXTRA_CB_OPTIONS% --no-test"
 )
 
+if NOT [%flow_run_id%] == [] (
+    set "EXTRA_CB_OPTIONS=%EXTRA_CB_OPTIONS% --extra-meta flow_run_id=%flow_run_id% remote_url=%remote_url% sha=%sha%"
+)
+
 call :end_group
 
 :: Build the recipe
 echo Building recipe
-conda.exe mambabuild "recipe" -m .ci_support\%CONFIG%.yaml --suppress-variables %EXTRA_CB_OPTIONS%
+conda.exe build "recipe" -m .ci_support\%CONFIG%.yaml --suppress-variables %EXTRA_CB_OPTIONS%
 if !errorlevel! neq 0 exit /b !errorlevel!
 
 :: Prepare some environment variables for the upload step

--- a/recipe/get_cpu_triplet.sh
+++ b/recipe/get_cpu_triplet.sh
@@ -1,0 +1,16 @@
+get_triplet() {
+  local CPU_ARCH
+  if [[ "$1" == "linux-64" ]]; then
+    CPU_ARCH="x86_64"
+  elif [[ "$1" == "linux-ppc64le" ]]; then
+    CPU_ARCH="powerpc64le"
+  elif [[ "$1" == "linux-aarch64" ]]; then
+    CPU_ARCH="aarch64"
+  elif [[ "$1" == "linux-s390x" ]]; then
+    CPU_ARCH="s390x"
+  else
+    echo "Unknown architecture"
+    exit 1
+  fi
+  echo $CPU_ARCH-conda-linux-gnu
+}

--- a/recipe/install_clang_symlinks.sh
+++ b/recipe/install_clang_symlinks.sh
@@ -5,3 +5,10 @@ maj_version="${PKG_VERSION%%.*}"
 ln -s "${PREFIX}/bin/clang-${maj_version}" "${PREFIX}/bin/clang-cl"
 ln -s "${PREFIX}/bin/clang-${maj_version}" "${PREFIX}/bin/clang-cpp"
 ln -s "${PREFIX}/bin/clang-${maj_version}" "${PREFIX}/bin/clang"
+
+if [[ "$target_platform" == "linux-"* ]]; then
+  source ${RECIPE_DIR}/get_cpu_triplet.sh
+  CHOST=$(get_triplet $target_platform)
+  ln -s "${PREFIX}/bin/clang-${maj_version}" "${PREFIX}/bin/${CHOST}-clang"
+  echo "--sysroot ${PREFIX}/${CHOST}/sysroot" >> ${PREFIX}/bin/${CHOST}-clang.cfg
+fi

--- a/recipe/install_clang_symlinks.sh
+++ b/recipe/install_clang_symlinks.sh
@@ -10,5 +10,6 @@ if [[ "$target_platform" == "linux-"* ]]; then
   source ${RECIPE_DIR}/get_cpu_triplet.sh
   CHOST=$(get_triplet $target_platform)
   ln -s "${PREFIX}/bin/clang-${maj_version}" "${PREFIX}/bin/${CHOST}-clang"
+  # for background, see comment in install_clangxx.sh
   echo "--sysroot ${PREFIX}/${CHOST}/sysroot" >> ${PREFIX}/bin/${CHOST}-clang.cfg
 fi

--- a/recipe/install_clangxx.sh
+++ b/recipe/install_clangxx.sh
@@ -6,3 +6,14 @@ ln -s $PREFIX/bin/clang  $PREFIX/bin/$HOST-clang++
 if [[ "$variant" == "hcc" ]]; then
   ln -s $PREFIX/bin/clang++  $PREFIX/bin/hcc
 fi
+
+if [[ "$target_platform" == "linux-"* ]]; then
+  source ${RECIPE_DIR}/get_cpu_triplet.sh
+  CHOST=$(get_triplet $target_platform)
+  ln -s "${PREFIX}/bin/clang++" "${PREFIX}/bin/${CHOST}-clang++"
+  # In cross compiling case we set CONDA_BUILD_SYSROOT to host platform
+  # which makes compiling for build platform not work correctly.
+  # This overrides CONDA_BUILD_SYSROOT so that native compilation
+  # works correctly in cross compilation scenarios.
+  echo "--sysroot ${PREFIX}/${CHOST}/sysroot" >> ${PREFIX}/bin/${CHOST}-clang++.cfg
+fi

--- a/recipe/install_clangxx.sh
+++ b/recipe/install_clangxx.sh
@@ -11,9 +11,11 @@ if [[ "$target_platform" == "linux-"* ]]; then
   source ${RECIPE_DIR}/get_cpu_triplet.sh
   CHOST=$(get_triplet $target_platform)
   ln -s "${PREFIX}/bin/clang++" "${PREFIX}/bin/${CHOST}-clang++"
-  # In cross compiling case we set CONDA_BUILD_SYSROOT to host platform
+  # In the cross compiling case, we set CONDA_BUILD_SYSROOT to host platform
   # which makes compiling for build platform not work correctly.
-  # This overrides CONDA_BUILD_SYSROOT so that native compilation
-  # works correctly in cross compilation scenarios.
+  # The following overrides CONDA_BUILD_SYSROOT, so that a clang for a given
+  # CHOST will always use the appropriate sysroot. In particular, this means
+  # that a clang in the build environment will work correctly for its native
+  # architecture also in cross compilation scenarios.
   echo "--sysroot ${PREFIX}/${CHOST}/sysroot" >> ${PREFIX}/bin/${CHOST}-clang++.cfg
 fi

--- a/recipe/install_clangxx.sh
+++ b/recipe/install_clangxx.sh
@@ -1,7 +1,6 @@
 #!/bin/bash
 
 ln -s $PREFIX/bin/clang  $PREFIX/bin/clang++
-ln -s $PREFIX/bin/clang  $PREFIX/bin/$HOST-clang++
 
 if [[ "$variant" == "hcc" ]]; then
   ln -s $PREFIX/bin/clang++  $PREFIX/bin/hcc

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -1,6 +1,6 @@
 {% set version = "17.0.4" %}
 {% set major_version = version.split(".")[0] %}
-{% set build_number = 0 %}
+{% set build_number = 1 %}
 
 {% set minor_aware_ext = major_version %}
 {% if version.split(".")[1] | int > 0 %}
@@ -392,7 +392,6 @@ outputs:
         - zstd     # [unix]
     requirements:
       host:
-        - gcc_impl_{{ target_platform }}  # [linux]
         - {{ pin_subpackage("clang-" ~ major_version, exact=True) }}
         # Use the same requirements as the top-level requirements
         - libcxx {{ cxx_compiler_version }}  # [osx]
@@ -401,7 +400,9 @@ outputs:
         - zlib
         - zstd
       run:
-        - gcc_impl_{{ target_platform }}  # [linux]
+        - binutils_impl_{{ target_platform }}  # [linux]
+        - sysroot_{{ target_platform }}        # [linux]
+        - libgcc-devel_{{ target_platform }}   # [linux]
         - {{ pin_subpackage("clang-" ~ major_version, exact=True) }}
     test:
       commands:
@@ -431,6 +432,7 @@ outputs:
         - zlib
         - zstd
       run:
+        - libstdcxx-devel_{{ target_platform }}  # [linux]
         - {{ pin_subpackage("clang", exact=True) }}
     test:
       requires:

